### PR TITLE
`flutter analyze` cleanup

### PR DESCRIPTION
**Pull request by Hixie**
* Make `--dartdocs` work in all modes.
* Make `analyze-sample-code.dart` more resilient.
* Add a test for `analyze-sample-code.dart`.
* Minor cleanup in related code and files.

### CODE REVIEW

**Comment by Hixie** [comment]
This fixes some regressions caused by https://github.com/flutter/flutter/pull/16979 that went unnoticed (basically, the dartdoc verifier's error reporting has been broken since then).

**Comment by Hixie** [comment]
cc @gspencergoog (this affects how dartdocs are analyzed)
cc @tvolkert (this affects the tool)
cc @devoncarew (this affects the analyzer)

File: dev/bots/analyze-sample-code.dart [hunk 1]
```diff
@@ -212,50 +219,40 @@ dependencies:
...
       workingDirectory: temp.path,
     );
-    stderr.addStream(process.stderr);
-    final List<String> errors = await process.stdout.transform<String>(utf8.decoder).transform<String>(const LineSplitter()).toList();
-    if (errors.first == 'Building flutter tool...')
+    final List<String> errors = <String>[];
+    errors.addAll(await process.stderr.transform<String>(utf8.decoder).transform<String>(const LineSplitter()).toList());
+    errors.addAll(await process.stdout.transform<String>(utf8.decoder).transform<String>(const LineSplitter()).toList());
+    if (errors.isNotEmpty && (errors.first.contains(' issues found. (ran in ') || errors.last.contains(' issue found. (ran in '))) {
+      errors.removeAt(0); // the "23 issues found" message goes onto stderr, which is concatenated first
```
Comment (tvolkert) [inline dev/bots/analyze-sample-code.dart]: But you use `errors.last` in the cast of a single issue found?

File: dev/bots/analyze-sample-code.dart [hunk 2]
```diff
@@ -212,50 +219,40 @@ dependencies:
...
+      if (errors.isNotEmpty && errors.last == '')
```
Comment (tvolkert) [inline dev/bots/analyze-sample-code.dart]: `errors.last.isEmpty`

File: dev/bots/analyze-sample-code.dart [hunk 3]
```diff
@@ -212,50 +219,40 @@ dependencies:
...
+        errors.removeLast();
+    }
+    if (errors.isNotEmpty && errors.first == 'Building flutter tool...')
       errors.removeAt(0);
-    if (errors.first.startsWith('Running "flutter packages get" in '))
+    if (errors.isNotEmpty && errors.first.startsWith('Running "flutter packages get" in '))
```
Comment (tvolkert) [inline dev/bots/analyze-sample-code.dart]: Might this not be the first entry in `errors`, e.g. if there were lines written to `stderr`?

Comment (tvolkert) [reply dev/bots/analyze-sample-code.dart]: (same issue with 'Building flutter tool...')

File: dev/bots/analyze-sample-code.dart [hunk 4]
```diff
@@ -212,50 +219,40 @@ dependencies:
...
       errors.removeAt(0);
     int errorCount = 0;
+    final String kBullet = Platform.isWindows ? ' - ' : ' • ';
+    final RegExp errorPattern = new RegExp('^ +([a-z]+)$kBullet(.+)$kBullet(.+):([0-9]+):([0-9]+)$kBullet([-a-z_]+)\$');
```
Comment (tvolkert) [inline dev/bots/analyze-sample-code.dart]: Are those leading spaces ever likely to be tabs?

Comment (tvolkert) [inline dev/bots/analyze-sample-code.dart]: Can the error code ever contain capital letters?

File: dev/bots/analyze-sample-code.dart [hunk 5]
```diff
@@ -99,15 +99,21 @@ class Section {
...
   final Directory temp = Directory.systemTemp.createTempSync('analyze_sample_code_');
   int exitCode = 1;
   bool keepMain = false;
   final List<String> buffer = <String>[];
   try {
     final File mainDart = new File(path.join(temp.path, 'main.dart'));
     final File pubSpec = new File(path.join(temp.path, 'pubspec.yaml'));
-    final Directory flutterPackage = new Directory(path.join(_flutterRoot, 'packages', 'flutter', 'lib'));
+    Directory flutterPackage;
+    if (arguments.length == 1) {
```
Comment (gspencergoog) [inline dev/bots/analyze-sample-code.dart]: How about using the args package here? Or at least do something simple like force it to be length 2, where the first arg has to be "--package".

That way, if we add more arguments someday, this one will be named, and not taking up the unnamed "first argument" spot, and the callers (probably) won't have to change how they call it.

File: dev/bots/analyze-sample-code.dart [hunk 6]
```diff
@@ -212,50 +219,40 @@ dependencies:
...
-      final int start = error.indexOf(kBullet);
-      final int end = error.indexOf(atRegExp);
-      if (start >= 0 && end >= 0) {
-        final String message = error.substring(start + kBullet.length, end);
-        final String atMatch = atRegExp.firstMatch(error)[0];
-        final int colon2 = error.indexOf(kColon, end + atMatch.length);
-        if (colon2 < 0) {
+      final Match parts = errorPattern.matchAsPrefix(error);
+      if (parts != null) {
+        final String message = parts[2];
```
Comment (gspencergoog) [inline dev/bots/analyze-sample-code.dart]: I really wish that Dart/JavaScript regexes had named groups...

**Review by gspencergoog — CHANGES_REQUESTED** [review]

File: dev/bots/analyze-sample-code.dart (hunk 1 above)
Comment (Hixie) [reply dev/bots/analyze-sample-code.dart]: oops

Comment (Hixie) [reply dev/bots/analyze-sample-code.dart]: fixed

File: dev/bots/analyze-sample-code.dart (hunk 2 above)
Comment (Hixie) [reply dev/bots/analyze-sample-code.dart]: fixed

File: dev/bots/analyze-sample-code.dart (hunk 3 above)
Comment (Hixie) [reply dev/bots/analyze-sample-code.dart]: the issues found line is the only one that goes to stderr, the rest are stdout.
I've added a null to separate the two and verified that after removing the one line, we see the null.

File: dev/bots/analyze-sample-code.dart (hunk 4 above)
Comment (Hixie) [reply dev/bots/analyze-sample-code.dart]: no, this is output from the analyzer, which does space-based indenting

Comment (Hixie) [reply dev/bots/analyze-sample-code.dart]: I don't think so, but for sanity I've just made the whole regexp case-insensitive.

**Review by tvolkert — APPROVED** [review]
LGTM

**Review by gspencergoog — APPROVED** [review]

**Comment by Hixie** [comment]
@fkorotkov Same here, the analyze shard and the tool_tests-linux shard just seem to die without notice half-way through.

**Comment by fkorotkov** [comment]
@Hixie very weird. As per our previous conversation I've made a change to log exit codes > 1 to see if there is something weird going on. Next time it happens we'll get more info. 🙌

**Pull request merged**